src/cliffordsynthesis/encoding/TableauEncoder.cpp

```diff
@@ -68,6 +82,48 @@ void TableauEncoder::assertTableau(const Tableau& tableau,
   lb->assertFormula(vars.r[t] == LogicTerm(targetR, n));
 }
 
+void TableauEncoder::assertMappingConstraints() const {
+  // if p_i_j is set column i is mapped to column j between 0 and 1
+  for (std::size_t i = 0U; i < N; ++i) {
+    for (std::size_t j = 0U; j < N; ++j) {
+      lb->assertFormula(
+          LogicTerm::implies(vars.p[i][j], vars.x[1][j] == vars.x[0][i]));
+      lb->assertFormula(
+          LogicTerm::implies(vars.p[i][j], vars.z[1][j] == vars.z[0][i]));
+    }
+  }
+  // assert that r vals are unchanges between 0 and 1
+  lb->assertFormula(LogicTerm::eq(vars.r[0], vars.r[1]));
+  // assert that for every i and j exactly one p variable is set
+  for (std::size_t i = 0U; i < N; ++i) {
+    const int32_t vr = 0;
+    const int32_t vr1 = 1;
+    LogicTerm sumRow = LogicTerm(vr);
+    for (std::size_t j = 0U; j < N; ++j) {
+      sumRow = sumRow + vars.p[i][j];
+    }
+    lb->assertFormula(sumRow == LogicTerm(vr1));
+    const int32_t vc = 0;
+    const int32_t vc1 = 1;
+    LogicTerm sumCol = LogicTerm(vc);
+    for (std::size_t j = 0U; j < N; ++j) {
+      sumCol = sumCol + vars.p[j][i];
+    }
+    lb->assertFormula(sumCol == LogicTerm(vc1));
```

Comment:
Instead of doing symbolic arithmetics here, consider using the `exactlyOneCmdr` function (also used in the `GateEncoder` class.